pkg/resource/deploy/step_executor.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/util/cancel"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+)
+
+const (
+	// Dummy workerID for synchronous operations.
+	synchronousWorkerID = -1
+
+	// Utility constant for easy debugging.
+	stepExecutorLogLevel = 4
+)
+
+// A Chain is a sequence of Steps that must be executed in the given order.
+type Chain = []Step
+
+// stepExecutor is the component of the engine responsible for taking steps and executing
+// them, possibly in parallel if requested. The step generator operates on the granularity
+// of "chains", which are sequences of steps that must be executed exactly in the given order.
+// Chains are a simplification of the full dependency graph DAG within Pulumi programs. Since
+// Pulumi language hosts can only invoke the resource monitor once all of their dependencies have
+// resolved, we (the engine) can assume that any chain given to us by the step generator is already
+// ready to execute.
+type stepExecutor struct {
+	plan        *Plan    // The plan currently being executed.
+	opts        Options  // The options for this current plan.
+	preview     bool     // Whether or not we are doing a preview.
+	pendingNews sync.Map // Resources that have been created but are pending a RegisterResourceOutputs.
+
+	workers        sync.WaitGroup // WaitGroup tracking the worker goroutines that are owned by this step executor.
+	incomingChains chan Chain     // Incoming chains that we are to execute
+
+	cancel *cancel.Source // Cancellation source for this step executor.
+
+	sawError uint32 // atomic boolean indicating whether or not the step excecutor saw that there was an error.
+}
+
+//
+// The stepExecutor communicates with a stepGenerator by listening to a channel. As the step generator
+// generates new chains that need to be executed, the step executor will listen to this channel to execute
+// those steps.
+//
+
+// Execute submits a Chain for asynchronous execution. The execution of the chain will begin as soon as there
+// is a worker available to execute it.
+func (se *stepExecutor) Execute(chain Chain) {
+	select {
```

Comment:
Hmmmmmmm... this looks supremely bizarre. Don't we actually need to run something here? I'm sure I'm just missing something important here. Could you add a comment about "why these blocks are empty"?

Reply:
I'll add a comment, but for posterity:

The only thing we actually need to do here is stuff the incoming chain onto `se.incomingChains`. However, since that chain is unbuffered, this can block if no workers are alive to receive the incoming chain (i.e. they just exited because they saw an error.)

The second arm of the select aborts the write to `se.incomingChains` if the step executor is shutting down, since it doesn't make sense to execute any more steps anyway.